src/sysext/sysext.c

```diff
@@ -262,9 +276,33 @@ static int unmerge_hierarchy(
                 if (r == 0)
                         break;
 
+                r = read_one_line_file(work_dir_info_file, &escaped_work_dir);
+                if (r < 0) {
+                        if (r != -ENOENT)
+                                return log_error_errno(r, "Failed to read '%s': %m", work_dir_info_file);
+                } else {
+                        ssize_t l;
+
+                        l = cunescape_length(escaped_work_dir, r, 0, &work_dir);
+                        if (l < 0)
+                                return log_error_errno(l, "Failed to unescape work directory path: %m");
+                }
+
+                r = umount_verbose(LOG_DEBUG, dot_dir, MNT_DETACH|UMOUNT_NOFOLLOW);
+                if (r < 0) {
+                        if (r != -EINVAL)
+                                return log_error_errno(r, "Failed to unmount '%s': %m", dot_dir);
+                }
+
                 r = umount_verbose(LOG_ERR, p, MNT_DETACH|UMOUNT_NOFOLLOW);
                 if (r < 0)
-                        return log_error_errno(r, "Failed to unmount file system '%s': %m", p);
+                        return r;
+
+                if (work_dir) {
+                        r = rm_rf(work_dir, REMOVE_ONLY_DIRECTORIES | REMOVE_ROOT | REMOVE_MISSING_OK | REMOVE_PHYSICAL);
```

Comment:
hmm, REMOVE_ONLY_DIRECTORIES?

Reply:
Yeah. From my observations, after unmounting the overlay, workdir has only one empty subdirectory named "work" and that's it. If there's something else then I suppose something went wrong and it's better to stop here, right?

Reply:
hmm, at the very least this needs a comment. but i'd just kill it all tbh.